text/0000-macro-at-most-once-rep.md

```diff
@@ -0,0 +1,174 @@
+- Feature Name: macro-at-most-once-rep
+- Start Date: 2018-01-17
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+
+Summary
+-------
+
+Add a repetition specifier to macros to repeat a pattern at most once: `$(pat)?`. Here, `?` behaves like `+` or `*` but represents at most one repetition of `pat`.
+
+Motivation
+----------
+
+There are two specific use cases in mind.
+
+## Macro rules with optional parts
```

Comment:
Could you point out some macros from crates.io that would benefit from this syntax for optional parts? The example given here strikes me as contrived.

Reply:
I would have used it in [this crate](https://crates.io/crates/guilt-by-association) to match the following syntaxes in one rule:

1. `trait Foo ...`
2. `trait Foo: Bar `
3. `pub trait Foo ...`
4. `pub trait Foo: Bar `

namely the rule:

```rust
($vis:vis trait $traitname:ident $(: $supertrait:ident)? ...)
```

Reply:
Damn, it's a local ambiguity. Strike that example...

Reply:
This also strikes me as a candidate for reimplementation: https://docs.rs/clap/2.29.4/clap/macro.clap_app.html